server/v2/appmanager/service.go

```diff
@@ -52,13 +58,40 @@ type AppManager[T transaction.Tx] struct {
 
 	db store.Store
 
+	mempool mempool.Mempool[T]
+
 	lastBlockHeight *atomic.Uint64
 
 	initGenesis func(ctx context.Context, genesisBytes []byte) error
 
+	prepareHandler appmanager.PrepareHandler[T]
+
 	stf *stf.STF[T]
 }
 
+// BuildBlock builds a block when requested by consensus. It will take in a list of transactions and return a list of transactions
+func (a AppManager[T]) BuildBlock(ctx context.Context, block appmanager.BlockRequest, totalSize uint32) ([]T, error) {
+
+	txs, err := a.mempool.GetTxs(ctx, totalSize)
+	if err != nil {
+		return nil, err
+	}
+
+	currentState, err := a.db.NewStateAt(block.Height)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create new state for height %d: %w", block.Height, err)
+	}
+
+	// run txs through handler
+	// TODO handle if optimistic or immediate execution
```

Comment:
handle optimistic  and/or immediate in consensus, prepare would call all functions needed in the app.